pkg/client/requestor.go

```diff
@@ -0,0 +1,36 @@
+/*
+Copyright (c) 2018 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package client contains the types and functions used to communicate with other services using
+// queues and topics.
+package client
+
+// ResponseCallback is called when a response to a request is received
+// m is the response message
+// requestID is the id of the request this message responds
+type ResponseCallback func(response Message, requestID string) error
+
+// RequestorSpec is a helper struct for building requestors.
+type RequestorSpec struct {
+	Destination   string
+	ResponseQueue string
```

Comment:
Is `destination` a queue? if so - can a Requestor send a request on Queue A and receive the response on Queue B?

Reply:
Actually a requestor *must* send the request using one queue and receive the response using another queue, otherwise the responder would get the response that it send itself. To make that clear I'd suggest to name these two fields `RequestsQueue` and `ResponsesQueue`.

In the future we could make the `ResponsesQueue` optional, and create them automatically, or use temporary queues, but for now I think it is better to just explicitly indicate the queue where you want to receive the responses.

Reply:
I agree `destination` is a bit confusing for me.
Another question: Does each `Requester/Responder` (on each side respectively) have a queue per *connection* or per *request/response cycle*? that is can `Requester` and `Responder` exchange a number of messages on a single connection between them or each time they interact have to exchange queues. In the case they exchange multiple messages on a single pair of queues - what if two messages are sent on the *same RequestQueue* and one takes longer to process than the other resulting in the responses coming back in different order to the `Requester` in the `ResponseQueue`?

Reply:
There is one requests queue that exists independently of the life-cycle of the requestors and responders. It is created in advance, same like a database table is created in advance. Both the requestor and responders should always use this queue for requests: it is their contract.

For the responses queue, it is up to the requestor; it can decide to use always the same queue, which is created in advance, or it can dynamically create it. I think that at this point it is simpler to create it in advance.

The order of the responses isn't relevant in this scenario: the requestor remembers that it has a set of requests pending, requests that haven't received a response yet. When the response is received it calls the corresponding response handle function. That may happen out of order: the response to second request may arrive before the response to first request; that is completely OK and we need to cope with it. In situations where the order is mportant, the requestor should avoid making the second request before it receives the response to the first request.

Reply:
:+1:  Great, Thanks for clearing that up.

Reply:
@jhernand, but currently we only allow one response per request? meaning if I get a response to request id X I remove this id from the pending requests?

We can later extend that to a protocol that can accept multiple responses but has to receive a FIN sort of message to note this is the last message and there is no more need to wait on pending requests

Reply:
Yes, we currently only have one response per request. And we will certainly want to extend it to multiple responses per request (client sends `GET` for all cluster, service sends multiple responses, each with a subset of clusters), so yes, we will eventually need that kind of FIN response.